firmware/thruna/BMS/src/io/io_imd.c

```diff
@@ -4,28 +4,33 @@
 #include "hw_pwmInput.h"
 #include "io_time.h"
 
-static PwmInput *pwm_input;
+static PwmInput pwm_input;
+
+uint8_t io_pwm_counter_tick(void)
+{
+    return hw_pwm_counter_tick();
+}
 
 void io_imd_init(const PwmInputConfig *pwm_input_config)
 {
-    io_pwmInput_init(pwm_input, pwm_input_config);
+    io_pwmInput_init(&pwm_input, pwm_input_config);
 }
 
 float io_imd_getFrequency(void)
 {
-    return hw_pwmInput_getFrequency(pwm_input);
+    return hw_pwmInput_getFrequency(&pwm_input);
 }
 
 float io_imd_getDutyCycle(void)
 {
-    return hw_pwmInput_getDutyCycle(pwm_input);
+    return hw_pwmInput_getDutyCycle(&pwm_input);
 }
 
 void io_imd_inputCaptureCallback(TIM_HandleTypeDef *htim)
 {
-    if (htim == pwm_input->config->htim)
```

Comment:
Perhaps move the timeout counter to here so that you don't have the HW not working issue. Have it reset anytime this input capture callback is triggered.